Add capture the fully qualified type name for raised exceptions in spans

emdneto opened the pull request:
Fixes #3824 

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] `tox -e py39-opentelemetry-sdk`

# Does This PR Require a Contrib Repo Change?

Answer the following question based on these examples of changes that would require a Contrib Repo Change:
- [The OTel specification](https://github.com/open-telemetry/opentelemetry-specification) has changed which prompted this PR to update the method interfaces of `opentelemetry-api/` or `opentelemetry-sdk/`
- The method interfaces of `test/util` have changed
- Scripts in `scripts/` that were copied over to the Contrib repo have changed
- Configuration files that were copied over to the Contrib repo have changed (when consistency between repositories is applicable) such as in
    - `pyproject.toml`
    - `isort.cfg`
    - `.flake8`
- When a new `.github/CODEOWNER` is added
- Major changes to project information, such as in:
    - `README.md`
    - `CONTRIBUTING.md`

- [ ] Yes. - Link to PR: 
- [X] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -994,9 +994,11 @@ def record_exception(
             )
-        )
+        )    
+        module = exception.__class__.__module__ if exception.__class__.__module__ != "builtins" else ""
```
`exception.__module__` should return the same value

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -994,9 +994,11 @@ def record_exception(
                 type(exception), value=exception, tb=exception.__traceback__
             )
-        )
+        )    
```
trailing whitespace, run `tox -e lint` too

xrmx reviewed (commented):
Would be nice to add tests to check that it is working fine

emdneto commented:
> Would be nice to add tests to check that it is working fine

@xrmx Thank you for your valuable review. I added a test and fixed the items you mentioned

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -995,8 +995,17 @@ def record_exception(
         )
+        module = (
+            exception.__module__
+            if exception.__class__.__module__ != "builtins"
```
you can drop `__class__` here too

emdneto replied:
If I drop it fails for builtins. Think we can use `type(exception).__module__`

xrmx commented on opentelemetry-sdk/tests/trace/test_trace.py:
```diff
@@ -1144,6 +1148,28 @@ def error_status_test(context):
+    def test_record_exception_fqn(self):
+        span = trace._Span("name", mock.Mock(spec=trace_api.SpanContext))
+        module_name = "dummy.module"
+        with mock.patch.object(DummyError, "__module__", module_name):
```
What do you get if you don't mock it? If it's stable there's no need to mock

emdneto replied:
Just to have a custom pathing in module name instead of `tests.trace.test_trace.DummyError` to assert.

emdneto replied:
I can also use `self.__module__` to get the module name to assert and remove the mock

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -995,8 +995,17 @@ def record_exception(
+            else ""
+        )
+        exc_type = (
+            f"{module}.{exception.__class__.__qualname__}"
```
What about using `type(exception).__qualname__` that is a bit shorter?

emdneto replied:
Agreed

xrmx commented on opentelemetry-sdk/tests/trace/test_trace.py:
```diff
@@ -1144,6 +1148,25 @@ def error_status_test(context):
+    def test_record_exception_fqn(self):
+        span = trace._Span("name", mock.Mock(spec=trace_api.SpanContext))
+        exception = DummyError("error")
+        exception_type = f"{exception.__module__}.{type(exception).__qualname__}"
```
This would be better to hardcode that to reimplement what the tested code is doing

emdneto replied:
Agreed. Let me put the full path

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -995,8 +995,18 @@ def record_exception(
                 type(exception), value=exception, tb=exception.__traceback__
             )
         )
+        module = (
```
i think using a single if would make it more readable instead of doing two inline ifs.What do you think?

emdneto replied:
Agreed. Let me improve it

tammy-baylis-swi commented:
Please could you also add a changelog line linking this PR.

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -995,8 +995,15 @@ def record_exception(
+        qualname = type(exception).__qualname__
+        exception_type = (
+            f"{module}.{qualname}"
+            if module and module != "builtins"
```
Why not use `builtins` as part of the qualified name?

xrmx replied:
The example on [specs](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/exception.md) uses `OSError` as an example and I think it'll be more natural for Python developers.

emdneto replied:
Mainly because of readability and this [statement](https://peps.python.org/pep-0737/#add-pytype-getfullyqualifiedname-function)

emdneto commented:
@lzchen any idea why auto-merge failed?
